lib/ui/payment/payment_web_page.dart

```diff
@@ -0,0 +1,249 @@
+                    }
+                  },
+                  onLoadError: (controller, navigationAction, code, msg) async {
+                    if (_dialog.isShowing()) {
+                      await _dialog.hide();
+                    }
+                  },
+                  onLoadHttpError: (controller, navigationAction, code, msg) async {
+                    _logger.info("onHttpError with $code and msg = $msg");
+                  },
+                  onLoadStop: (controller, navigationAction) async {
+                    _logger.info("loadStart" + navigationAction.toString());
+                    if (_dialog.isShowing()) {
+                      await _dialog.hide();
+                    }
+                  },
+                ),
+              ),
+            ].where((Object o) => o != null).toList(),
+          ),
+        ));
+  }
+
+  @override
+  void dispose() {
+    _dialog.hide();
+    super.dispose();
+  }
+
+  // show dialog to handle accidental back press.
+  Future<bool> _buildPageExitWidget(BuildContext context) {
+    return showDialog(
+        context: context,
+        builder: (context) => AlertDialog(
+                title: Text('Are you sure you want to exit?'),
+                actions: <Widget>[
+                  TextButton(
+                      child: Text('yes'),
+                      onPressed: () => Navigator.of(context).pop(true)),
+                  TextButton(
+                      child: Text('no'),
+                      onPressed: () => Navigator.of(context).pop(false)),
+                ]));
+  }
+
+  bool _isPaymentActionComplete(Uri loadingUri) {
+    return loadingUri.toString().startsWith(kWebPaymentRedirectUrl);
+  }
+
+  Future<void> handlePaymentResponse(Uri uri) async {
+    var queryParams = uri.queryParameters;
+    var paymentStatus = uri.queryParameters['status'] ?? '';
+    _logger.fine('handle payment response with status $paymentStatus');
+    if (paymentStatus == 'success') {
+      await _handlePaymentSuccess(queryParams);
+    } else if ('fail' == paymentStatus) {
+      var reason = queryParams['reason'] ?? '';
+      await _handlePaymentFailure(reason);
+    } else {
+      // should never reach here
+      _logger.severe("unexpected status", uri.toString());
+       showGenericErrorDialog(context);
+    }
+  }
+
+  Future<void> _handlePaymentFailure(String reason) async {
+    await showDialog(
+        context: context,
+        barrierDismissible: false,
+        builder: (context) => AlertDialog(
+                title: Text('payment failed'),
+                content:
+                    Text("unfortunately your payment failed due to $reason"),
+                actions: <Widget>[
+                  TextButton(
+                      child: Text('ok'),
+                      onPressed: () {
+                        Navigator.of(context).pop('dialog');
+                      }),
+                ]));
+    Navigator.of(context).pop(true);
```

Comment:
Does this not pop the dialog?

Reply:
This will take user to back page. Err dialog is popped on 'ok' press.